File: documentation-examples-java/src/java/dev/typr/foundations/docs/analysis/QueryAnalysisNamed.java
package dev.typr.foundations.docs.analysis;

import dev.typr.foundations.Fragment;
import dev.typr.foundations.PgTypes;
import dev.typr.foundations.RowCodec;
import dev.typr.foundations.QueryAnalysis;
import dev.typr.foundations.QueryAnalyzer;
import java.sql.Connection;

@SuppressWarnings("unused")
public class QueryAnalysisNamed {
    record User(int id, String name, String email) {}

    private final Connection connection = null; // placeholder
    private final int userId = 1;

    private final RowCodec<User> userRowCodec =
        RowCodec.<User>builder()
            .field(PgTypes.int4, User::id)
            .field(PgTypes.text, User::name)
            .field(PgTypes.text, User::email)
            .build(User::new);

    //start
    void analyzeNamedQuery() {
        var query =
            Fragment.of("""
                    SELECT id, name, email
                    FROM users WHERE id =
                    """)
                .value(PgTypes.int4, userId)
                .query(userRowCodec.all())
                .named("findUserById");

        // The name shows up in the error report
        QueryAnalysis analysis =
            QueryAnalyzer.analyze(query, connection).getFirst();

        if (!analysis.succeeded()) {
            throw new AssertionError(analysis.report());
        }
    }
    //stop
}

File: documentation-examples-scala/src/scala/dev/typr/foundationssc/docs/analysis/QueryAnalysisNamed.scala
package dev.typr.foundationssc.docs.analysis
import dev.typr.foundationssc.*
import dev.typr.foundationssc.Fragment.*
import dev.typr.foundationssc.data.*

@SuppressWarnings(Array("unused"))
object QueryAnalysisNamed:
  case class User(id: Int, name: String, email: String)

  private val userId = 1

  private val userRowCodec: RowCodec[User] = RowCodec
    .builder[User]()
    .field(PgTypes.int4)(_.id)
    .field(PgTypes.text)(_.name)
    .field(PgTypes.text)(_.email)
    .build(User.apply)

  // start
  def analyzeNamedQuery(connection: Connection): Unit =
    val query =
      sql"""SELECT id, name, email
            FROM users
            WHERE id = ${PgTypes.int4(userId)}"""
        .query(userRowCodec.all())
        .named("findUserById")

    // The name shows up in the error report
    val analysis =
      QueryAnalyzer.analyze(query, connection).head

    if !analysis.succeeded() then throw AssertionError(analysis.report())
  // stop
